python/ml4ir/model/losses/listwise_losses.py

```diff
@@ -8,4 +12,39 @@ def _make_loss_fn(self, **kwargs):
         Additionally can pass in record positions to handle positional bias
 
         """
-        raise NotImplementedError
+        bce = losses.BinaryCrossentropy(reduction=Reduction.SUM)
+        mask = kwargs.get("mask")
+
+        def _loss_fn(y_true, y_pred):
+            batch_size = tf.cast(tf.shape(y_true)[0], tf.float32)
+
+            # Mask the padded records
+            y_true = tf.gather_nd(y_true, tf.where(tf.equal(mask, tf.constant(1.0))))
+            y_pred = tf.gather_nd(y_pred, tf.where(tf.equal(mask, tf.constant(1.0))))
+
+            # Reshape the tensors
+            y_true = tf.expand_dims(tf.squeeze(y_true), axis=-1)
```

Comment:
what does `expand_dims(squeeze(y_true), axis=-1)` do to the shape?

Reply:
`tf.squeeze()` is probably an overkill here as `y_true` and `y_pred`(if we have written a good model architecture) should already be 1-dimensional tensor.

Otherwise,
`tf.squeeze()` collapses empty dimensions on the tensor, so **[1, x, 1] -> [x]**
`tf.expand_dims(axis=-1)` is similar to `np.new_axis` in numpy, it changes a tensor of dimension **[x] -> [x, 1]**

Reply:
sounds reasonable.